Decoder Only Generation Support

jonathangomesselman opened the pull request:
## Description

In this PR we focus on adding generation support for Decoder Only models. In doing so, we also re-structure parts of the generation workflow. Below is an executive summary of the changes added!

The primary differences between generation for EncoderDecoder and DecoderOnly models are:
1. The df column / sample data used for generation
2. The generation processing for each sample (i.e. the logic previously in the `utils.seq2seq.generation.generate_sample_output` function).

Thankfully, similar to the processing of model logprob outputs, after sample level generation / processing, logic can be shared around token alignment and the overall saving of information to the dataframe - since in the end we expect the same information for generated data.

**The main structural change is to address (2) from above.**

To do this we leverage the `BaseSeq2SeqDataFormatter` class and define a new function `generate_sample` that is is over-written by `EncoderDecoderDataFormatter` and `DecoderOnlyDataFormatter`. Since generation occurs within `Seq2SeqDataLogger`, when we call `generation.add_generated_output_to_df` (from `Seq2SeqDataLogger.add_generated_output_to_df`), we now pass along the correct `formatter`. Then within `generation.generate_on_batch`, in place of calling `generate_sample_output` we call the respective `formatter.generate_sample`. 

The above description gives a high level intuition for the big changes in code structure. Additionally, there are some new parameters needed for generation with DecoderOnly models, such as the `sample_id` we are generating on (necessary to help isolate just the "input prompt" prompt to the model by removing the response tokens). 

**Addressing Difference (1)**
This is a fairly simple but important difference. 
- EncoderDecoder models can directly use the `Input` text column for generation.
- DecoderOnly models need the `formatted_prompt` text column to account for additional prompt formatting applied to the `Input` / `Target` pair.

**Additional Things to Note**
When first testing, I noticed that I was having a strange off by 1 error in the alignment. Essentially, the top-5 tokens for each "word" were actually the top-5 tokens for the next word. 

The problem ended up being that the tokenizer I was using added a `<bos>` token by default. However, for DecoderOnly models, although I am aligning just the generated tokens / string, I do so "as if" it were directly pre-fixed by the prompt. Therefore, I do not want any `<bos>` tokens added because I do not have logits for these, and thus the token alignments were shifted by one!

If the above does not immediately make much sense that is okay. The main takeaway is that within `utils.seq2seq.generation.generate_on_batch`, I explicitly decode and tokenize without adding *or* removing special tokens - i.e. I use exactly the generated ids - to avoid alignment issues. Additionally, this issue pointed me to potential challenges if tokenizers add <bos> tokens that we should test for when adding test completeness.

## Demonstration

You should now be able to generate using the `Seq2Seq_Decoder_Only_Dq_Fake_Data` notebook.

Here is an example [run](https://console.dev.rungalileo.io/insights/8de23502-d6cc-44a5-8452-0999ff6d6c92/cf1b9db4-fa8f-436f-8f2c-4b4af55406c5?dataframeColumns=bleu%3B%26data_error_potential%3B%26generated_output%3B%26generated_token_logprobs%3B%26generated_uncertainty%3B%26input%3B%26perplexity%3B%26rouge%3B%26target&groupedBy=galileo_input_length&taskType=8).

bogdan-galileo commented on dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:
```diff
@@ -318,10 +323,29 @@ def add_generated_output_to_df(
+            split_key=split,
         )
+        # The formatted_prompts column is no longer needed
+        df = df.drop([S2SIC.formatted_prompts])
```
Fixing it 
![Screenshot 2023-12-06 at 4 11 44 PM](https://github.com/rungalileo/dataquality/assets/112427971/a316cabe-cd57-4bed-8f73-09b1cc640179)

jonathangomesselman commented:
> It takes an unreasonable amount of time (almost 10 minutes !!!) on ![Screenshot 2023-12-06 at 4 32 37 PM](https://private-user-images.githubusercontent.com/112427971/288553823-b322ae57-b812-4734-8e9e-0b280e42818c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE5NTc5ODEsIm5iZiI6MTcwMTk1NzY4MSwicGF0aCI6Ii8xMTI0Mjc5NzEvMjg4NTUzODIzLWIzMjJhZTU3LWI4MTItNDczNC04ZTllLTBiMjgwZTQyODE4Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwN1QxNDAxMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NjlhODM0OGY4NzIxOTEzYWMxYmNlNTc2N2NkNGI0MTgzNjM1MTUzYmU1NDkwMTVhMWZlNGI3Njk1ZDI3ZDEwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.WmjOFD2HGVnwGmwUP58hfYV2_O0Sqhuiijsz9XNPK_E)
> 
> Tested end to end and it does generate
> 
> * enc_dec run https://console.dev.rungalileo.io/insights/e24eb571-8eaf-45a1-beb0-86c6488f4f1a/314f813f-8548-4b34-ab67-0e5e6b402c5c?split=validation&taskType=8
> * dec_only run https://console.dev.rungalileo.io/insights/e24eb571-8eaf-45a1-beb0-86c6488f4f1a/a30025a0-2526-42d8-a903-dabbe2f9b66f?split=training&taskType=8

@bogdan-galileo - Are you using just the sample notebook I created?

It certainly seems odd. I ran many local tests and for 10 samples with the facebook `opt-170m` model, it takes:
- max_tokens generated = 15 --> `55s` total
- max_tokens generated = 100 --> `42s` for generation and `111s` total

See this [run](https://console.dev.rungalileo.io/insights/8de23502-d6cc-44a5-8452-0999ff6d6c92/e87441ce-684d-426e-a042-accc7c5b505d?dataframeColumns=data_error_potential%3B%26generated_token_logprobs%3B%26input%3B%26perplexity%3B%26target&groupedBy=galileo_input_length&taskType=8) for `max_tokens generated = 100`.

bogdan-galileo commented:
> > It takes an unreasonable amount of time (almost 10 minutes !!!) on ![Screenshot 2023-12-06 at 4 32 37 PM](https://private-user-images.githubusercontent.com/112427971/288553823-b322ae57-b812-4734-8e9e-0b280e42818c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE5NTc5ODEsIm5iZiI6MTcwMTk1NzY4MSwicGF0aCI6Ii8xMTI0Mjc5NzEvMjg4NTUzODIzLWIzMjJhZTU3LWI4MTItNDczNC04ZTllLTBiMjgwZTQyODE4Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwN1QxNDAxMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NjlhODM0OGY4NzIxOTEzYWMxYmNlNTc2N2NkNGI0MTgzNjM1MTUzYmU1NDkwMTVhMWZlNGI3Njk1ZDI3ZDEwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.WmjOFD2HGVnwGmwUP58hfYV2_O0Sqhuiijsz9XNPK_E)
> > Tested end to end and it does generate
> > 
> > * enc_dec run https://console.dev.rungalileo.io/insights/e24eb571-8eaf-45a1-beb0-86c6488f4f1a/314f813f-8548-4b34-ab67-0e5e6b402c5c?split=validation&taskType=8
> > * dec_only run https://console.dev.rungalileo.io/insights/e24eb571-8eaf-45a1-beb0-86c6488f4f1a/a30025a0-2526-42d8-a903-dabbe2f9b66f?split=training&taskType=8
> 
> @bogdan-galileo - Are you using just the sample notebook I created?
> 
> It certainly seems odd. I ran many local tests and for 10 samples with the facebook `opt-170m` model, it takes:
> 
> * max_tokens generated = 15 --> `55s` total
> * max_tokens generated = 100 --> `42s` for generation and `111s` total
> 
> See this [run](https://console.dev.rungalileo.io/insights/8de23502-d6cc-44a5-8452-0999ff6d6c92/e87441ce-684d-426e-a042-accc7c5b505d?dataframeColumns=data_error_potential%3B%26generated_token_logprobs%3B%26input%3B%26perplexity%3B%26target&groupedBy=galileo_input_length&taskType=8) for `max_tokens generated = 100`.

I used the notebook in `rungalileo/examples` for encoder-decoder and the one in the dq repo for decoder-only. But it seems to take a long time after dq work is done, which is why I'm so confused. Plat has been updated how runners job are queued and all, maybe it's related ?

bogdan-galileo commented on dataquality/schemas/seq2seq.py:
```diff
@@ -38,6 +38,10 @@ class Seq2SeqInputCols(str, Enum):
     system_prompts = "system_prompts"
 
 
+class Seq2SeqInputTempCols(str, Enum):
```
@jonathangomesselman @elboy3 I created this class because this is a temporary column that will not be saved. If we put it in `Seq2SeqInputCols` we have the attribute test that fails. Let me know what you think

jonathangomesselman replied:
Definitely an @elboy3 question!

elboy3 replied:
Works for me. Nit:  I would just call it `Seq2SeqTemporaryCols` and remove input and we can use it for input or output dfs

elboy3 commented on dataquality/loggers/data_logger/seq2seq/formatters.py:
```diff
@@ -71,6 +81,73 @@ def format_text(
+        input_str: str
+            Input string context used to seed the generation
+        tokenizer: PreTrainedTokenizerFast
+        max_input_tokens: the max number of tokens to use for tokenization
```
```suggestion
        max_input_tokens: int
            the max number of tokens to use for tokenization
```

elboy3 commented on dataquality/loggers/data_logger/seq2seq/formatters.py:
```diff
@@ -71,6 +81,73 @@ def format_text(
+        model: PreTrainedModel,
+        max_input_tokens: int,
+        generation_config: GenerationConfig,
+        input_id: Optional[int] = None,
+        split_key: Optional[str] = None,
```
these are missing from docstring

elboy3 commented on dataquality/loggers/data_logger/seq2seq/formatters.py:
```diff
@@ -71,6 +81,73 @@ def format_text(
+        generation_config: GenerationConfig
+            Users generation config specifying the parameters for generation
+
+        Return:
+        -------
+        model_generation: ModelGeneration
+            - generated_ids: np.ndarray of shape - [seq_len]
+            - generated_token_logprobs: np.ndarray of shape - [seq_len]
+            - generated_top_logprobs: List[List[Tuple[str, float]]]
+        """
```
nice!

elboy3 commented on dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:
```diff
@@ -88,14 +90,11 @@ def __init__(self, meta: Optional[MetasType] = None) -> None:
+        # Only required for Decoder-Only models
         self.formatted_prompts: List[str] = []
         # Formatter distinguishes behavior between EncoderDecoder and DecoderOnly
-        from dataquality.loggers.data_logger.seq2seq.formatters import (
-            BaseSeq2SeqDataFormatter,
-        )
```
very nice, love removing import  from here

jonathangomesselman replied:
Yeah I think it was necessary haha

elboy3 commented on dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:
```diff
@@ -290,22 +298,21 @@ def convert_large_string(self, df: DataFrame) -> DataFrame:
         """
-        logger_config = cls.logger_config
-        if split not in logger_config.generation_splits:
+        # logger_config = cls.logger_config
```
🧹

elboy3 commented on dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:
```diff
@@ -318,10 +325,27 @@ def add_generated_output_to_df(
             )
         assert isinstance(max_input_tokens, int)
 
+        assert generation_config is not None
```
this could be none if you watch without the config but pass in generation splits as not none? maybe we can do that validation in watch (if we don't already)

jonathangomesselman replied:
Oh I see! But wait could the formatter ever be None?

elboy3 commented on dataquality/utils/seq2seq/decoder_only.py:
```diff
@@ -26,42 +26,55 @@ def extract_tokenized_responses(
+        np.array(tokenized_formatted_prompt) == response_template[0]
+    )[0]
+    response_token_ids_start_idx = None
+    for i in range(len(matched_indices)):
+        match_idx = matched_indices[-(i + 1)]
```
could we do some kidn of `for idx in reverse(matched_indices)` ?

jonathangomesselman replied:
Yeah 100% - I didn't really think about this!

elboy3 commented on dataquality/loggers/data_logger/seq2seq/formatters.py:
```diff
@@ -338,6 +457,63 @@ def format_text(
+    ) -> ModelGeneration:
+        """Generate response for a single sample - Decoder Only"""
+
+        # Retrieve the length of the response tokens
+        assert split_key, "Decoder-Only models require a valid `split_key`"
+        assert input_id is not None, (
+            "Decoder-Only models require the `sample_id` "
+            "that is being used for generation"
+        )
```
Not asking for a change just noting - i feel like these assertion errors should guide the user on what they should fix. If these assertions fail, what went wrong?

jonathangomesselman replied:
Since this is not really something the user would mess up should we just have no error message? This is mostly for typing

elboy3 commented on dataquality/loggers/data_logger/seq2seq/formatters.py:
```diff
@@ -338,6 +457,63 @@ def format_text(
+            verbose=False,
+            return_tensors="pt",
+        )["input_ids"]
+        # TODO Watch for empty targets!
```
what's the TODO action here?

jonathangomesselman replied:
Oh I think because [:, :-0] would be bad, since it would give none of the ids - but we actually want the whole thing! I will fix this.

jonathangomesselman replied:
I added a fix that should handle this case seamlessly

elboy3 commented on dataquality/loggers/data_logger/seq2seq/formatters.py:
```diff
@@ -338,6 +457,63 @@ def format_text(
+            return_tensors="pt",
+        )["input_ids"]
+        # TODO Watch for empty targets!
+        just_prompt_ids = formatted_prompt_ids[:, :-num_response_labels].to(
```
just_prompt_ids feels like an odd var name.

elboy3 replied:
`prompt_ids_without_response` ? maybe not better 🤷‍♀️

elboy3 reviewed (approved):
Looking good. I'm approving but I think there's some cleanup (on prints, assertion statements, and comments) that could be done. 

Stellar!

jonathangomesselman commented on dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:
```diff
@@ -280,7 +283,12 @@ def convert_large_string(self, df: DataFrame) -> DataFrame:
         """
         df_copy = df.copy()
-        for text_col in [S2SIC.input.value, S2SIC.target.value]:
+        # TODO Include formatted prompt?
```
Wait what is this function doing? When did you add this?

jonathangomesselman commented:
Thanks for all the great comments @elboy3 @bogdan-galileo - I noticed on or two bugs that must have snuck in with other cleanup changes. So I also have fixed those. Let's get this in next week!
